agent/connect/ca/provider_vault.go

```diff
@@ -551,13 +551,39 @@ func (v *VaultProvider) GenerateIntermediate() (string, error) {
 	}
 
 	// Set the intermediate backend to use the new certificate.
-	_, err = v.writeNamespaced(v.config.IntermediatePKINamespace, v.config.IntermediatePKIPath+"intermediate/set-signed", map[string]interface{}{
+	importResp, err := v.writeNamespaced(v.config.IntermediatePKINamespace, v.config.IntermediatePKIPath+"intermediate/set-signed", map[string]interface{}{
 		"certificate": intermediate.Data["certificate"],
 	})
 	if err != nil {
 		return "", err
 	}
 
+	if importResp != nil {
+		// Assume we're on Vault 1.11+. We'll assume we only have one issuer
+		// with a key.
+		mapping := importResp.Data["mapping"].(map[string]string)
+		var intermediateId string
+		for issuer, key := range mapping {
+			if key != "" {
+				intermediateId = issuer
+
+				// Could be safe and check:
+				// importedIssuers := importResp.Data["imported_issuers"].([]string)
+				// and make sure this newly imported issuer is in the set of new unique
+				// issuers.
```

Comment:
I think this is unnecessary since Consul will always generate a new UUID for intermediate CSR. Correct me if I'm wrong

Reply:
Hmm, no this is a Vault generated/returned UUID.

Vault's new import APIs (in 1.11, they've all been unified -- so `/config/ca` and `/intermediate/set-signed` all function the same) return the following data structure:

```
imported_issuers: []string
imported_keys: []string
mapping: dict(issuer->key)
```

So if you post say:

```
Cert A
Key for Cert A
Cert B
Key for Cert B
```

and nothing exists, you'd end up with:

```
imported_issuers: A, B
imported_keys: A, B
mapping:
  A->A
  B->B
```

But if you import:

```
Cert A
Cert B
Key B
```

and the key for A already exists in the mount, you'd get:

```
imported_issuers: A, B
imported_keys: B
mapping:
  A->A
  B->B
```

i.e, the key wasn't present on import, but was present on the mount. 

---

^^ This is what we want to guard against, with the above comment. It shouldn't happen, but it is good for data safety.

In particular, if Consul posts `{Root Cert, Intermediate Cert}` and someone posted the Root Cert's key into this mount, your mapping would look like the above (minus the `imported_keys` since you didn't import a key -- it was already internal on generation).

So, we need to ensure that only the new issuer gets selected as default, and not the root cert. However, if the root cert was net new (someone previously imported its key, but didn't import the root itself), this method would fail. It seems unlikely to me, tbh.


Alternatively, on the `POST /intermediate/generate/internal` call, you could save the `key_id` response and ensure that this newly imported/selected issuer is the one matching that existing `key_id`. This is safer and what I'd prefer if you don't like the above.

---

Similarly, if you do:

```
Key A
Cert B
Key for B
```

you'd get:

```
imported_issuers: B
imported_keys: A, B
mapping:
  B->B
```

because, while the key was imported and new, you wouldn't see the cert in the mapping.

What we want to guard against.